SymbolDisplay: Show underlying non-default type for enums

MaStr11 opened the pull request:
Fixes #52490

![GmzWNndhpn](https://user-images.githubusercontent.com/24472128/114724700-20dec100-9d3c-11eb-9b75-c2adc8e2a8f7.gif)

MaStr11 commented on src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Types.cs:
```diff
@@ -406,6 +407,17 @@ private void AddNameAndTypeArgumentsOrParameters(INamedTypeSymbol symbol)
 
+        private void AddUnderlyingEnumType(INamedTypeSymbol symbol)
+        {
+            if (symbol.TypeKind == TypeKind.Enum && symbol.EnumUnderlyingType.SpecialType != SpecialType.System_Int32)
```
Are there any format options to consider?

MaStr11 commented on src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Types.cs:
```diff
@@ -406,6 +407,17 @@ private void AddNameAndTypeArgumentsOrParameters(INamedTypeSymbol symbol)
+                AddSpace();
+                AddPunctuation(SyntaxKind.ColonToken);
+                AddSpace();
+                AddSpecialTypeKeyword(symbol.EnumUnderlyingType);
```
Maybe better like so?
```suggestion
                VisitNamedType(symbol.EnumUnderlyingType);
```

MaStr11 commented on src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Types.cs:
```diff
@@ -395,6 +395,7 @@ private void AddNameAndTypeArgumentsOrParameters(INamedTypeSymbol symbol)
                 AddDelegateParameters(symbol);
             }
 
+            AddUnderlyingEnumType(symbol);
```
Adding this here assumes, that the underlying enum type is treated more or less like type parameters.

CyrusNajmabadi commented on src/Compilers/CSharp/Test/Symbol/SymbolDisplay/SymbolDisplayTests.cs:
```diff
@@ -7736,5 +7736,75 @@ class A {
+                SymbolDisplayPartKind.Keyword,
+                SymbolDisplayPartKind.Space,
+                SymbolDisplayPartKind.EnumName);
+        }
```
so this should only be shown in the case where we're checking on the top level type.  if the enum is referenced anywhere (like in a method signature) it woudl not be shown there.  Can you write tests showing that that's the case.

Also, we'd likely need this for VB.

MaStr11 replied:
> so this should only be shown in the case where we're checking on the top level type

Yes. Adding the underlying type unconditionaly breaks all kinds of features (CI build failures confirm this). I haven't found an obvious way to prevent this. I think we need a new `formatOption`. But which one and where exactly should it go?

If such an option is present I could turn it of before any `VisitNamedType` calls e.g. if the SymbolDisplay is requested for an EnumMember.

Can you please give me some hints about what exactly you would expect here to solve this?

CyrusNajmabadi replied:
> Yes. Adding the underlying type unconditionaly breaks all kinds of features

I think we should only doing this in the codepth that is also adding `enum`.  So we only would show: `enum E : byte`  not `E : byte` just because `E` was used in a reference location.  The SymbolDisplay code should already be partitioned to support this concept afaict.

MaStr11 commented:
@CyrusNajmabadi Ready for review. I started from scratch and moved away from `SymbolDisplayVisitor`. I added the logic to `AbstractSymbolDisplayService.AbstractSymbolDescriptionBuilder` instead, which I felt is the correct place.

CyrusNajmabadi commented:
That works for me.

CyrusNajmabadi commented:
LGTM.  Thanks!
